File: src/main/java/gargoyle/calendar/util/load/DefaultLoaders.java
package gargoyle.calendar.util.load;

import gargoyle.calendar.util.resources.Resource;

import javax.imageio.ImageIO;
import java.applet.Applet;
import java.applet.AudioClip;
import java.awt.*;
import java.awt.image.BufferedImage;
import java.io.IOException;
import java.io.InputStream;
import java.net.URL;
import java.text.MessageFormat;

public final class DefaultLoaders extends Loaders {
    public static final DefaultLoaders INSTANCE = new DefaultLoaders();

    private DefaultLoaders() {
        addLoader(Font.class, DefaultLoaders::loadFont);
        addLoader(BufferedImage.class, DefaultLoaders::loadImage);
        addLoader(Image.class, DefaultLoaders::loadImage);
        addLoader(AudioClip.class, DefaultLoaders::loadAudioClip);
    }


    private static AudioClip loadAudioClip(Resource resource) {
        URL url = resource.getUrl();
        return Applet.newAudioClip(url);
    }


    private static Font loadFont(Resource resource) throws IOException {
        try (InputStream stream = resource.getInputStream()) {
            return Font.createFont(Font.TRUETYPE_FONT, stream);
        } catch (FontFormatException e) {
            throw new IOException(MessageFormat.format("cannot load font from {0}", resource), e);
        }
    }


    private static BufferedImage loadImage(Resource resource) throws IOException {
        try (InputStream stream = resource.getInputStream()) {
            return ImageIO.read(stream);
        }
    }
}

File: cal-core/src/main/kotlin/gargoyle/calendar/util/load/DefaultLoaders.kt
@file:Suppress("DEPRECATION")

package gargoyle.calendar.util.load

import gargoyle.calendar.util.resources.Resource
import java.applet.Applet
import java.applet.AudioClip
import java.awt.Font
import java.awt.Image
import java.awt.image.BufferedImage
import javax.imageio.ImageIO

object DefaultLoaders : Loaders() {
    init {
        addLoader(Font::class) { loadFont(it) }
        addLoader(BufferedImage::class) { loadImage(it) }
        addLoader(Image::class) { loadImage(it) }
        addLoader(AudioClip::class) { loadAudioClip(it) }
    }

    private fun loadAudioClip(resource: Resource): AudioClip =
        Applet.newAudioClip(resource.url)

    private fun loadFont(resource: Resource): Font =
        run {
            resource.inputStream.use { stream -> Font.createFont(Font.TRUETYPE_FONT, stream) }
        }

    private fun loadImage(resource: Resource): BufferedImage =
        resource.inputStream.use { stream -> requireNotNull(ImageIO.read(stream)) { "cannot read image from $resource" } }
}
